networkx/utils/decorators.py

```diff
@@ -392,84 +355,868 @@ def random_array(dims, random_state=1):
+        Function whose random_state_argument is converted to a Random instance.
 
     Examples
     --------
     Decorate functions like this::
 
-       @py_random_state(0)
+       @py_random_state("random_state")
        def random_float(random_state=None):
            return random_state.rand()
 
+       @py_random_state(0)
+       def random_float(rng=None):
+           return rng.rand()
+
        @py_random_state(1)
-       def random_array(dims, random_state=1):
-           return random_state.rand(*dims)
+       def random_array(dims, seed=12345):
+           return seed.rand(*dims)
 
     See Also
     --------
     np_random_state
     """
 
-    @decorator
-    def _random_state(func, *args, **kwargs):
-        # Parse the decorator arguments.
-        try:
-            random_state_arg = args[random_state_index]
-        except TypeError as e:
-            raise nx.NetworkXError("random_state_index must be an integer") from e
-        except IndexError as e:
-            raise nx.NetworkXError("random_state_index is incorrect") from e
+    return argmap(create_py_random_state, random_state_argument)
+
+
+class argmap:
+    """A decorator to apply a map to arguments before calling the function
+
+    This class provides a decorator that maps (transforms) arguments of the function
+    before the function is called. Thus for example, we have similar code
+    in many functions to determine whether an argument is the number of nodes
+    to be created, or a list of nodes to be handled. The decorator provides
+    the code to accept either -- transforming the indicated argument into a
+    list of nodes before the actual function is called.
+
+    This decorator class allows us to process single or multiple arguments.
+    Or to process arguments that e.g. might need to be closed (or not) depending
+    on how they are input (used for file handle arguments that may either be
+    a string naming the file or an existing file handle). If the decorator opens
+    the file it should close it when done. Otherwise it should not close the file.
+    The arguments to be processed can be specified by string, naming the argument,
+    or by index, specifying the item in the args list.
+
+    Parameters
+    ----------
+    func : callable
+        The function to apply to arguments
+
+    *params: iterable of (int, str or tuple)
+        A list of parameters, specified either as strings (their names), ints
+        (numerical indices) or arbitrarily nested tuples thereof.
+
+    Examples
+    --------
+    Let's start with a caveat. Most of these examples will use `@argmap(...)`
+    to apply the decorator immediately to the function defined on the next line.
+    In the NetworkX codebase however, `argmap` is used within a function to
+    construct a decorator. That is, the decorator defines a mapping function
+    and then uses `argmap` to build the decorator from the mapping.
```

Comment:
I wasn't 100% sure what this meant.  Is this a correct phrasing?

The decorator calls the original function, after transforming the input arguments.